### PR TITLE
Paginated Delete for Directory & deleteRecursively

### PR DESCRIPTION
**Pull request by ibrahimrabab**
# Description

STG90 feature for paginated delete for DataLake directories. Paginated delete is only possible for service versions 2023-08-03 and above, and only when using OAuth for authentication.

Also adding `DataLakeDirectoryClient.deleteRecursively()`, which allows users to delete directories and all contents within directories.

### CODE REVIEW

File: sdk/storage/azure-storage-file-datalake/CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,8 @@
 ## 12.17.0-beta.1 (Unreleased)
 
 ### Features Added
+- Added `DataLakeDirectoryClient.deleteRecursively()` which allows users to delete a directory and all of its contents recursively. 
+- Added support for paginated directory delete when using AAD authentication. Note that this feature only applies to HNS storage accounts.
```
Comment (seanmcc-msft) [inline sdk/storage/azure-storage-file-datalake/CHANGELOG.md:7]:
```suggestion
- Added support for paginated directory delete when using AAD authentication. Note that this feature only applies to HNS storage accounts and when using token authentication.
```

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java [hunk 2]
```diff
@@ -284,6 +287,10 @@ CpkInfo getCpkInfo() {
         return this.customerProvidedKey;
     }
 
+    boolean isTokenCredentialAuthenticated() {
```
Comment (seanmcc-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java:290]: This method is not publicly accessible, correct?

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/DirectoryApiTests.java [hunk 3]
```diff
@@ -585,6 +601,46 @@ public void deleteDirDoesNotExistAnymore() {
         assertEquals(BlobErrorCode.BLOB_NOT_FOUND.toString(), e.getErrorCode());
     }
 
+    @Disabled("Requires manual OAuth setup and creates 5000+ files")
+    @DisabledIf("olderThan20230803ServiceVersion")
+    @Test
+    public void deletePaginatedDirectory() {
```
Comment (seanmcc-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/DirectoryApiTests.java:607]: Have we been able to successfully manually run this test?

Comment (ibrahimrabab) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/DirectoryApiTests.java:607]: Yes, confirmed that it continues to delete the directory successfully with a response code of 200, and confirmed the continuation token passed in the request and response headers.

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java (hunk 2 above)
Comment (ibrahimrabab) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java:290]: This is not a public method. It is protected within DataLakePathAsyncClient and cannot be accessed publicly.

**Review by seanmcc-msft — APPROVED** [review]

File: sdk/storage/azure-storage-file-datalake/CHANGELOG.md [hunk 4]
```diff
@@ -3,6 +3,8 @@
 ## 12.17.0-beta.1 (Unreleased)
 
 ### Features Added
+- Added `DataLakeDirectoryClient.deleteRecursively()` which allows users to delete a directory and all of its contents recursively. 
+- Added support for paginated directory delete when using AAD authentication. Note that this feature only applies to HNS storage accounts and when using token authentication.
```
Comment (srnagar) [inline sdk/storage/azure-storage-file-datalake/CHANGELOG.md:7]: It might be worth adding a link to documentation that explains what HNS is.

**Review by srnagar — APPROVED** [review]

File: sdk/storage/azure-storage-file-datalake/CHANGELOG.md (hunk 4 above)
Comment (ibrahimrabab) [reply sdk/storage/azure-storage-file-datalake/CHANGELOG.md:7]: Will add it when merging to main.

**Pull request merged**